pkg/codec/property_extractor.go

```diff
@@ -0,0 +1,98 @@
+package codec
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+)
+
+// NewPropertyExtractor creates a modifier that will extract a single property from a struct.
+// This modifier is lossy, as TransformToOffchain will discard unwanted struct properties and
+// return a single element. Calling TransformToOnchain will result in unset properties.
+func NewPropertyExtractor(fieldName string) Modifier {
+	m := &propertyExtractor{
+		modifierBase: modifierBase[bool]{
+			fields:           map[string]bool{fieldName: false},
+			onToOffChainType: map[reflect.Type]reflect.Type{},
+			offToOnChainType: map[reflect.Type]reflect.Type{},
+		},
+		fieldName: fieldName,
+	}
+
+	m.modifyFieldForInput = func(_ string, field *reflect.StructField, _ string, _ bool) error {
+		return nil
+	}
+
+	return m
+}
+
+type propertyExtractor struct {
+	modifierBase[bool]
+	fieldName string
+}
+
+func (e *propertyExtractor) TransformToOnChain(offChainValue any, _ string) (any, error) {
+	parts := strings.Split(e.fieldName, ".")
+	fieldName := parts[len(parts)-1]
+
+	fields := []reflect.StructField{
+		{Name: fieldName, Type: reflect.TypeOf(offChainValue)},
+	}
+
+	onChainStruct := reflect.StructOf(fields)
+	output := reflect.New(onChainStruct)
+	iOutput := reflect.Indirect(output)
+	iOutput.FieldByName(fieldName).Set(reflect.ValueOf(offChainValue))
+
+	return iOutput.Interface(), nil
+}
+
+func (e *propertyExtractor) TransformToOffChain(onChainValue any, _ string) (any, error) {
+	rItem := reflect.ValueOf(onChainValue)
+
+	_, ok := e.onToOffChainType[rItem.Type()]
+	if !ok {
+		return reflect.Value{}, fmt.Errorf("%w: cannot retype %v", types.ErrInvalidType, rItem.Type())
+	}
+
+	switch rItem.Kind() {
+	case reflect.Struct:
+	case reflect.Pointer:
+		rItem = rItem.Elem()
+		if rItem.Kind() != reflect.Struct {
```

Comment:
Today, you can do all the Modifiers on lists and get back a list of modified items.  It would be nice to allow that in all modifiers.

Ie: If you had []struct{A : string, B: int} and you extract A you should get back a []string with values from A

Reply:
Oh that's an interesting case. I can add that one in.

Reply:
done